pkg/image/controller/scheduled_image_controller_test.go

```diff
@@ -124,3 +124,124 @@ func TestScheduledImport(t *testing.T) {
 		t.Fatalf("should have not added item to scheduler: %#v", sched.scheduler)
 	}
 }
+
+func TestNeedsScheduling(t *testing.T) {
+	one := int64(1)
+	tests := map[string]struct {
+		is             *imagev1.ImageStream
+		expectedResult bool
+	}{
+		"none scheduled": {
+			is: &imagev1.ImageStream{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test", Namespace: "other", UID: "1", ResourceVersion: "1",
+					Annotations: map[string]string{imagev1.DockerImageRepositoryCheckAnnotation: "done"},
+					Generation:  1,
+				},
+				Spec: imagev1.ImageStreamSpec{
+					Tags: []imagev1.TagReference{
+						{
+							Name:         "default",
+							From:         &corev1.ObjectReference{Kind: "DockerImage", Name: "mysql:latest"},
+							Generation:   &one,
+							ImportPolicy: imagev1.TagImportPolicy{},
+						},
+					},
+				},
+				Status: imagev1.ImageStreamStatus{
+					Tags: []imagev1.NamedTagEventList{
+						{
+							Tag:   "default",
+							Items: []imagev1.TagEvent{{Generation: 1}},
+						},
+					},
+				},
+			},
+			expectedResult: false,
+		},
+		"sha ref not imported": {
+			is: &imagev1.ImageStream{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test", Namespace: "other", UID: "1", ResourceVersion: "1",
+					Annotations: map[string]string{imagev1.DockerImageRepositoryCheckAnnotation: "done"},
+					Generation:  1,
+				},
+				Spec: imagev1.ImageStreamSpec{
+					Tags: []imagev1.TagReference{
+						{
+							Name:         "default",
+							From:         &corev1.ObjectReference{Kind: "DockerImage", Name: "abc@sha256:3c87c572822935df60f0f5d3665bd376841a7fcfeb806b5f212de6a00e9a7b25"},
+							Generation:   &one,
+							ImportPolicy: imagev1.TagImportPolicy{Scheduled: true},
+						},
+					},
+				},
+				Status: imagev1.ImageStreamStatus{},
+			},
+			expectedResult: true,
+		},
+		"sha ref imported": {
+			is: &imagev1.ImageStream{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test", Namespace: "other", UID: "1", ResourceVersion: "1",
+					Annotations: map[string]string{imagev1.DockerImageRepositoryCheckAnnotation: "done"},
+					Generation:  1,
+				},
+				Spec: imagev1.ImageStreamSpec{
+					Tags: []imagev1.TagReference{
+						{
+							Name:         "default",
+							From:         &corev1.ObjectReference{Kind: "DockerImage", Name: "abc@sha256:3c87c572822935df60f0f5d3665bd376841a7fcfeb806b5f212de6a00e9a7b25"},
+							Generation:   &one,
```

Comment:
this test would be better if the spec generation was newer than the status generation, to confirm we still don't import a SHA tag that is already successfully imported, right?

Reply:
ok on further study of the code, these values matching is our signal that the import is done/successful.  But i'm still wondering how the spec.generation value gets updated...i'm wondering if someone updates the imagestream, whether spec.generation gets updated and that triggers another round of imports.  If we can confirm that's not the case, then i think it's reasonable to assume that "once imported successfully, spec.generation will never exceed spec.status, for a sha based tag, as long as our schedule import avoidance logic is sound"

Reply:
Ok it looks like the logic which says a tagRef has changed are here:  https://github.com/openshift/openshift-apiserver/blob/master/pkg/image/apiserver/registry/imagestream/strategy.go#L375-L423

That logic above then dictates whether the tagRef.generation is updated at https://github.com/openshift/openshift-apiserver/blob/master/pkg/image/apiserver/registry/imagestream/strategy.go#L217-L228

I believe all the conditions at https://github.com/openshift/openshift-apiserver/blob/master/pkg/image/apiserver/registry/imagestream/strategy.go#L375-L423 would be valid reasons for re-importing in our case, but let me know if you disagree @bparees

Reply:
yup, i agree, thanks for digging into it @gabemontero

Reply:
cool thanks